### PR TITLE
Add a progress bar to sync

### PR DESCRIPTION
**Pull request by Pobega**
The progress bar is mostly a rough estimate of how much work there is.
It checks the OS dirty pages before copying, compares it to the OS dirty
pages after the copy and uses that as its baseline for completion to
build a progress bar.

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -13,3 +13,4 @@ default-run = "writedisk"
 [dependencies]
 progress = "0.2"
 structopt = "0.3"
+procfs = "0.8.0"
```
Comment (nicholasbishop) [inline Cargo.toml:16]: nitpick: change the version to "0.8" and sort alphabetically

Comment (Pobega) [reply Cargo.toml]: 952e1a3

File: src/bin/wd_copier.rs [hunk 2]
```diff
@@ -44,8 +51,30 @@ fn main() {
...
 
-    println!("syncing...");
+    progress_bar.set_job_title("syncing... (2/2)");
+
+    let (tx, rx) = mpsc::channel();
+
+    let meminfo = procfs::Meminfo::new().unwrap();
+    let dirty_after_copy = meminfo.dirty - starting_dirty;
+
+    thread::spawn(move || loop {
```
Comment (nicholasbishop) [inline src/bin/wd_copier.rs:61]: If it's not too much of a pain, can the contents of this closure be moved to a separate function?

Comment (Pobega) [reply src/bin/wd_copier.rs:61]: Sure, give me a heads up when you've looked at everything else and I can handle that last. I wanna make sure the base concept here is acceptable

File: src/bin/wd_copier.rs [hunk 3]
```diff
@@ -15,7 +18,11 @@ struct Opt {
 fn main() {
     let opt = Opt::from_args();
 
+    let meminfo = procfs::Meminfo::new().unwrap();
+    let starting_dirty = meminfo.dirty;
```
Comment (nicholasbishop) [inline src/bin/wd_copier.rs:22]: Since the meminfo stuff is not critical to functionality, I'd rather it not use unwrap. Maybe add a new function that just returns 0 if `Meminfo::new` fails:

```rust
fn get_dirty_bytes() -> u64 { ... }
```

(I'm not sure if the return value is in bytes, so maybe this function name is wrong)

File: src/bin/wd_copier.rs (hunk 2 above)
Comment (nicholasbishop) [reply src/bin/wd_copier.rs:61]: I make no promise that I won't request other changes :) The concept seems acceptable though.

File: src/bin/wd_copier.rs (hunk 3 above)
Comment (Pobega) [reply src/bin/wd_copier.rs:22]: I think that makes sense, and it makes the scopes much clearer than just having a bunch of Object::new constructors called inline in main.

Comment (nicholasbishop) [reply src/bin/wd_copier.rs:22]: Hmm, shouldn't the `starting` value be calculated after the copy operation? Otherwise what is it measuring?

Comment (Pobega) [reply src/bin/wd_copier.rs:22]: It calculates based on the amount of dirty bytes before and after the copy, using that as a baseline.

My first implementation didn't have the earlier `starting_dirty` and seemed to work OK, but I'd think this might be more accurate on a currently active system.

Comment (nicholasbishop) [reply src/bin/wd_copier.rs:22]: Ah right, that makes sense

Comment (Pobega) [reply src/bin/wd_copier.rs]: dbc2118

File: src/bin/wd_copier.rs (hunk 2 above)
Comment (Pobega) [reply src/bin/wd_copier.rs]: a562f67

File: src/bin/wd_copier.rs [hunk 4]
```diff
@@ -12,10 +15,40 @@ struct Opt {
     dst: PathBuf,
 }
 
+fn get_dirty_bytes() -> u64 {
```
Comment (nicholasbishop) [inline src/bin/wd_copier.rs:18]: based on running `cat /proc/meminfo`, I think this is actually kB rather than bytes

File: src/bin/wd_copier.rs [hunk 5]
```diff
@@ -12,10 +15,40 @@ struct Opt {
...
+    let meminfo = procfs::Meminfo::new().unwrap();
```
Comment (nicholasbishop) [inline src/bin/wd_copier.rs:19]: This error should be handled instead of unwrapping

File: src/bin/wd_copier.rs [hunk 6]
```diff
@@ -12,10 +15,40 @@ struct Opt {
...
+            - ((get_dirty_bytes().saturating_sub(starting_dirty))
+                / (dirty_after_copy / 100));
+        progress_bar.reach_percent(percent as i32);
+        thread::sleep(Duration::from_millis(500));
+        match rx.try_recv() {
+            Ok(_) | Err(mpsc::TryRecvError::Disconnected) => {
+                break;
+            }
+            Err(mpsc::TryRecvError::Empty) => {}
+        }
```
Comment (nicholasbishop) [inline src/bin/wd_copier.rs:41]: I think this match could be simplified, maybe:

```rust
if matches!(rx.try_recv(), Ok(_) | Err(mpsc::TryRecvError::Disconnected) {
  return;
}
```

File: src/bin/wd_copier.rs [hunk 7]
```diff
@@ -12,10 +15,40 @@ struct Opt {
     dst: PathBuf,
 }
 
+fn get_dirty_bytes() -> u64 {
+    let meminfo = procfs::Meminfo::new().unwrap();
+    meminfo.dirty
+}
+
+fn sync_progress_bar(
```
Comment (nicholasbishop) [inline src/bin/wd_copier.rs:23]: docstring describing what this func is doing would be good

File: src/bin/wd_copier.rs (hunk 4 above)
Comment (Pobega) [reply src/bin/wd_copier.rs:18]: The file is but the interface isn't; `/proc/meminfo` shows `Cached:          4905980 kB`, but `println!("{}", procfs::Meminfo::new().unwrap().cached);` shows `5037039616` which is apparently that value reported by bytes.

Also keep in mind it's not actually kB, it's KiB (https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/6/html/deployment_guide/s2-proc-meminfo) but it's left written as kB due to legacy concerns.

File: src/bin/wd_copier.rs [hunk 8]
```diff
@@ -12,10 +15,40 @@ struct Opt {
...
+    rx: mpsc::Receiver<()>,
+    mut progress_bar: progress::Bar,
+    starting_dirty: u64,
+    dirty_after_copy: u64,
+) {
+    progress_bar.set_job_title("syncing... (2/2)");
+    loop {
+        let percent = 100
+            - ((get_dirty_bytes().saturating_sub(starting_dirty))
+                / (dirty_after_copy / 100));
```
Comment (nicholasbishop) [inline src/bin/wd_copier.rs:33]: This expression is a bit complicated to reason about. Maybe add a new `calc_percentage` function that takes a min-max range and the current value.

File: src/bin/wd_copier.rs (hunk 4 above)
Comment (nicholasbishop) [reply src/bin/wd_copier.rs:18]: Good info. Looks good

File: src/bin/wd_copier.rs (hunk 6 above)
Comment (Pobega) [reply src/bin/wd_copier.rs:68]: Good call, I didn't know about `matches!`

78ecdc8

File: src/bin/wd_copier.rs (hunk 7 above)
Comment (Pobega) [reply src/bin/wd_copier.rs:50]: Is this the right way to do it? 464bf28

File: src/bin/wd_copier.rs (hunk 8 above)
Comment (Pobega) [reply src/bin/wd_copier.rs]: b63e642

File: src/bin/wd_copier.rs (hunk 5 above)
Comment (Pobega) [reply src/bin/wd_copier.rs]: Handled in b63e642. Additionally in the error case it now skips the progress bar altogether and falls back to the old `println!`

File: src/bin/wd_copier.rs (hunk 7 above)
Comment (nicholasbishop) [reply src/bin/wd_copier.rs:50]: This can be a little less formal -- it's not a library so we're not generating html docs. I would reword the docstring to talk more about what it's doing -- using the dirty value from `/proc/meminfo` to show a progress bar that estimates the time remaining for the sync to complete. You don't need to explicitly mention the arguments, especially `progress_bar` which is self explanatory.

File: src/bin/wd_copier.rs [hunk 9]
```diff
@@ -12,10 +15,67 @@ struct Opt {
...
+}
+
+/// Calculates the percentage of max as it approaches min
+///
+/// # Arguments
+///
+/// * `current` - current value of progress
+/// * `min` - target; percent = 100 when current <= min
+/// * `max` - starting point; percent = 0 when current == max
+fn calc_percent(current: u64, min: u64, max: u64) -> u64 {
```
Comment (nicholasbishop) [inline src/bin/wd_copier.rs:33]: You could consider using a https://doc.rust-lang.org/std/ops/struct.RangeInclusive.html for min and max.

File: src/bin/wd_copier.rs (hunk 7 above)
Comment (nicholasbishop) [reply src/bin/wd_copier.rs:50]: As a general rule if my function args are getting complex I repack them into a struct. For example you could have a `struct DirtyProgress` with `dirty_before_copy` and `dirty_after_copy` fields, and also have a method for getting the current estimated percent progress. I'm not sure if it's worth doing that here, but throwing it out as a possibility.

File: src/bin/wd_copier.rs [hunk 10]
```diff
@@ -12,10 +15,67 @@ struct Opt {
...
+    // Subtract min from current but clamp to 0u64
+    let numerator = current.saturating_sub(min);
+    let denominator = max / 100;
+    100 - (numerator / denominator)
```
Comment (nicholasbishop) [inline src/bin/wd_copier.rs:31]: potential division by zero here I think

File: src/bin/wd_copier.rs (hunk 7 above)
Comment (Pobega) [reply src/bin/wd_copier.rs:42]: How is this for the docstring? 0a98bc1

Comment (Pobega) [reply src/bin/wd_copier.rs:42]: Err, updated again in e5a078d since I realized that the docstring was implying the function will create the thread.

File: src/bin/wd_copier.rs (hunk 10 above)
Comment (Pobega) [reply src/bin/wd_copier.rs:31]: I think you're technically correct but the outer portion covers this scenario (due to the new error handling); https://github.com/Pobega/writedisk/blob/e5a078dcf9e903ab2f89c962fa3730357d1e57d2/src/bin/wd_copier.rs#L103

Comment (Pobega) [reply src/bin/wd_copier.rs:31]: Either way, is there an equivalent to `saturating_div` or a preferred way to handle this case? I don't think it's crazy to consider handling it explicitly even if the outer caller does.

Comment (Pobega) [reply src/bin/wd_copier.rs:31]: https://docs.rs/num/0.2.0/num/trait.CheckedDiv.html seems to be the most reasonable approach?

Comment (nicholasbishop) [reply src/bin/wd_copier.rs:31]: Since you're dealing with integers here, I would think `if denominator == 0 { ... }` would be the most straightforward way to handle it. Although as I look at it, I'm not sure we should be dealing with ints here. I guess it probably works out since the numbers will be fairly high...

Comment (Pobega) [reply src/bin/wd_copier.rs:31]: The current solution I have locally is this:

```
-    let denominator = max / 100;
-    100 - (numerator / denominator)
+//    let denominator = max / 100;
+    let denominator = 0;
+    100 - match numerator.checked_div(denominator) {
+        Some(result) => result,
+        None => 50,
+    }
```

I could go with either solution since I'm pretty impartial to it, just let me know which way you prefer.

Comment (Pobega) [reply src/bin/wd_copier.rs:34]: 8fb13e7

File: src/bin/wd_copier.rs (hunk 9 above)
Comment (Pobega) [reply src/bin/wd_copier.rs:27]: Is there an actual advantage to using this over `u64`? Mostly asking since I'm not using an iterator in this function, you end up just changing it to

```
fn calc_percent(current: u64, range: RangeInclusive<u64>) -> u64 {
     // Subtract min from current but clamp to 0u64
     let numerator = current.saturating_sub(range.start());                                                                                      
     let denominator = range.end() / 100;
```
Unless there's something I'm missing

Comment (nicholasbishop) [reply src/bin/wd_copier.rs:27]: Yep, the advantage just to make it hard to pass parameters in the wrong order. I like to avoid having multiple parameters of the same type next to each other, for example you could easily pass in `min, max, current` and not notice that it's supposed to be `current, min, max`. By using a range for the min/max, you make it obvious at the call site that the order is correct.
